scipy/stats/_rvs_sampling.py

```diff
@@ -168,3 +171,373 @@ def rvs_ratio_uniforms(pdf, umax, vmin, vmax, size=1, c=0, random_state=None):
+            d = qmc_engine.d if d is None else d
+        elif qmc_engine is None:
+            d = 1 if d is None else d
+            qmc_engine = stats.qmc.Halton(d)
+        else:
+            message = ("`qmc_engine` must be an instance of "
+                       "`scipy.stats.qmc.QMCEngine` or `None`.")
+            raise ValueError(message)
+
+        # `rvs` is flexible about whether `size` is an int or tuple, so this
+        # should be, too.
+        try:
+            tuple_size = tuple(size)
+        except TypeError:
+            tuple_size = (size,)
+
+        # Get uniform QRVS from qmc_random and transform it
+        uniform = qmc_engine.random(np.prod(tuple_size) or 1)
+        qrvs = self.ppf(uniform)
+
+        # Output reshaping for user convenience
+        if size is None:
+            return qrvs.squeeze()[()]
+        else:
+            if d == 1:
+                return qrvs.reshape(tuple_size)
+            else:
+                return qrvs.reshape(tuple_size + (d,))
+
+
+def _fni_input_validation(dist, tol, max_intervals):
+    """
+    Input validation for _fast_numerical_inverse.
+
+    """
+    if int(max_intervals) != max_intervals or max_intervals <= 1:
+        raise ValueError("`max_intervals' must be an integer greater than 1.")
+
+    tol = float(tol)  # if there's an exception, raise it now
+
+    return dist, tol, max_intervals
+
+
+def _fast_numerical_inverse(dist, tol=1e-12, max_intervals=100000):
+    """
+    Generate fast, approximate PPF (inverse CDF) of probability distribution.
+
+    `_fast_numerical_inverse` accepts `dist`, a frozen instance of
+    `scipy.stats.rv_continuous`, and returns a object `fni` with methods
+    that approximate `dist.ppf` and `dist.rvs`. For some distributions,
+    these methods may be faster than those of `dist` itself.
+
+    Parameters
+    ----------
+    dist : scipy.stats.rv_frozen
+        Frozen distribution for which fast approximate PPF is desired.
+    tol : float, optional
+        u-error tolerance. The default is 1e-12.
+    max_intervals : int, optional
+        Maximum number of intervals in the cubic Hermite Spline used to
+        approximate the percent point function. The default is 100000.
+
+    Returns
+    -------
+    H: scipy.interpolate.CubicHermiteSpline
+        Interpolant of the distributions's PPF.
+    intervals : int
+        The number of intervals of the interpolant.
+    midpoint_error : float
+        The maximum u-error at an interpolant interval midpoint.
+
+    """
+    dist, tol, max_intervals = _fni_input_validation(dist, tol, max_intervals)
+
+    # [1] Section 2.1: "For distributions with unbounded domain, we have to
+    # chop off its tails at [a] and [b] such that F(a) and 1-F(b) are small
+    # compared to the maximal tolerated approximation error."
+    p = np.array([dist.ppf(tol/10), dist.isf(tol/10)])  # initial interval
+
+    # [1] Section 2.3: "We then halve this interval recursively until
```

Comment:
Out of curiosity, did you try to replace this halving search by an optimization algorithm? Would it make sense to allow someone to plug another search method here?

Reply:
It is refining the grid by adding mesh points where the interpolating spline is not accurate according to the u-error: the round trip dist.cdf-FastNumericalInvese.ppf error. I'm not sure I see the connection with optimizers. I can imagine optimizing the grid points to reduce some measure of error; that would be a valid - but pretty different - algorithm.

Reply:
Sure sure it would be a totally different algo. The way I understood this is that you are trying to find a minimum for `eu` by changing `u`. Hence using a classical optimizer might help improve the performance. Maybe you saw this in the literature?

Reply:
No, not searching for a minimum. Trying to evaluate the CDF at enough points for the PPF to have good accuracy. Yes, algorithm is all in [1]. That's where the name `FastNumericalInverse` comes from, too. I'd recommend a look if you have time; i seem to remember it being an interesting read.

Reply:
Never mind I think I had it wrong since the beginning! Your comment about mesh refinement made me realise that I initially read this as we where just tracking a single set of parameters, but not updating a mesh.

I will go read this article yes, sorry about that.